pkg/ingester/ingester_v2_backfill.go

```diff
@@ -0,0 +1,499 @@
+}
+
+func newBackfillTSDB(userID string, backfillAge time.Duration, metrics *ingesterMetrics) *backfillTSDB {
+	return &backfillTSDB{
+		userID:      userID,
+		backfillAge: backfillAge,
+		metrics:     metrics,
+	}
+}
+
+func (b *backfillTSDB) appender(
+	createNewTSDB func(userID, dbDir string, minBlockDuration, maxBlockDuration int64, reg *prometheus.Registry) (*userTSDB, error),
+	backfillBlocksDir func(userID string) string,
+) *backfillAppender {
+	return &backfillAppender{
+		createNewTSDB:     createNewTSDB,
+		backfillBlocksDir: backfillBlocksDir,
+		backfillTSDB:      b,
+	}
+}
+
+// backfillAppender is an appender to ingest old data.
+// This _does not_ implement storage.Appender interface.
+// The methods of this appender should not be called concurrently.
+type backfillAppender struct {
+	backfillTSDB                 *backfillTSDB
+	newerAppender, olderAppender *backfillRangeAppender
+
+	// Keeping this methods of Ingester here to avoid the appender having a reference to the ingester.
+	createNewTSDB func(userID, dbDir string, minBlockDuration, maxBlockDuration int64, reg *prometheus.Registry) (*userTSDB, error)
+	// We are keeping the method instead of string for backfill dir because if there is a surge in backfill writes,
+	// it will create a new string for every push while we want it only to open a new TSDB.
+	backfillBlocksDir func(userID string) string
+}
+
+type backfillRangeAppender struct {
+	app        storage.Appender
+	db         *userTSDB
+	start, end int64
+}
+
+func newBackfillRangeAppender(db *backfillTSDBWrapper) *backfillRangeAppender {
+	return &backfillRangeAppender{
+		app:   db.db.Appender(),
+		db:    db.db,
+		start: db.start,
+		end:   db.end,
+	}
+}
+
+// add requires the samples to be within the backfill age.
+// Note: This method should not be called concurrently.
+func (a *backfillAppender) add(la []client.LabelAdapter, s client.Sample) (err error) {
+	app, db, err := a.getAppender(s)
+	if err != nil {
+		return err
+	}
+	startAppend := time.Now()
+
+	cachedRef, cachedRefExists := db.refCache.Ref(startAppend, client.FromLabelAdaptersToLabels(la))
+	// If the cached reference exists, we try to use it.
+	if cachedRefExists {
+		err = app.AddFast(cachedRef, s.TimestampMs, s.Value)
+		if err == nil || errors.Cause(err) != storage.ErrNotFound {
+			return err
+		}
+	}
+
+	// Copy the label set because both TSDB and the cache may retain it.
+	copiedLabels := client.FromLabelAdaptersToLabelsWithCopy(la)
+	ref, err := app.Add(copiedLabels, s.TimestampMs, s.Value)
+	if err == nil {
+		db.refCache.SetRef(startAppend, copiedLabels, ref)
+	}
+
+	return err
+}
+
+func (a *backfillAppender) getAppender(s client.Sample) (storage.Appender, *userTSDB, error) {
+	if !a.backfillTSDB.isWithinBackfillAge(s.TimestampMs) {
+		return nil, nil, storage.ErrOutOfBounds
+	}
+
+	// Fast path to not take the lock.
+	if a.newerAppender != nil && s.TimestampMs >= a.newerAppender.start && s.TimestampMs < a.newerAppender.end {
```

Comment:
Is it safe to access `a.newerAppender` and `a.olderAppender`, given their references may be replaced any time (and when they're replaced, a lock is used)?

Reply:
Yes, it's safe. Note that it is 1 appender per `v2Push` call and not shared across multiple push. The idea here is that the goroutine which is getting the appender will use it sequentially (which is the case in `v2Push`, each sample is iterated sequentially). So, while the DBs can be started in parallel which is protected by the lock, it is safe to access `newerAppender` and `olderAppender` without lock as they are the part of the appender and not the backfillTSDB.